File: python_backend/app/utils/document_parser.py
from __future__ import annotations

import asyncio
import os
from pathlib import Path

from openai import OpenAI
from pypdf import PdfReader
from unstructured.partition.auto import partition

ALLOWED_EXTS = {".pdf", ".docx", ".txt", ".mp3", ".mp4"}


def _openai_client() -> OpenAI | None:
    api_key = os.getenv("OPENAI_API_KEY")
    return OpenAI(api_key=api_key) if api_key else None


def _extract_with_unstructured(file_path: str) -> str:
    strategy = os.getenv("UNSTRUCTURED_STRATEGY", "hi_res")
    elements = partition(filename=file_path, strategy=strategy)
    text = "\n".join((getattr(item, "text", "") or "").strip() for item in elements).strip()
    if not text:
        raise ValueError("Unstructured returned empty text.")
    return text


def _extract_pdf_fast_no_ocr(file_path: str) -> str:
    reader = PdfReader(file_path)
    chunks: list[str] = []
    for page in reader.pages:
        chunks.append((page.extract_text() or "").strip())
    text = "\n".join(x for x in chunks if x).strip()
    if not text:
        raise ValueError("Fast PDF text extraction returned empty text.")
    return text


def _ensure_pdf_text_quality(text: str, filename: str) -> str:
    min_chars = int(os.getenv("PDF_MIN_TEXT_CHARS", "300"))
    if len((text or "").strip()) < min_chars:
        raise ValueError(
            f'PDF text extraction low quality for {filename}. '
            "Please upload a searchable PDF (not scanned image PDF)."
        )
    return text


async def _transcribe_audio(file_path: str) -> str:
    client = _openai_client()
    if not client:
        raise ValueError("OPENAI_API_KEY not configured for audio transcription")

    def _run() -> str:
        with open(file_path, "rb") as f:
            result = client.audio.transcriptions.create(model="whisper-1", file=f)
        return (getattr(result, "text", "") or "").strip()

    text = await asyncio.to_thread(_run)
    if not text:
        raise ValueError("Audio transcription returned empty text.")
    return text


async def extract_text_from_file(file_path: str, filename: str) -> str:
    ext = Path(filename).suffix.lower()
    if ext not in ALLOWED_EXTS:
        raise ValueError(f"Unsupported file type: {ext}. Allowed: PDF, DOCX, TXT, MP3, MP4")

    if ext == ".pdf":
        # First try a fast non-OCR extractor to avoid Tesseract dependency for digital PDFs.
        try:
            fast_text = await asyncio.to_thread(_extract_pdf_fast_no_ocr, file_path)
            if fast_text:
                return _ensure_pdf_text_quality(fast_text, filename)
        except Exception:
            pass

        # Fall back to Unstructured (may require OCR deps for scanned PDFs).
        extracted = await asyncio.to_thread(_extract_with_unstructured, file_path)
        return _ensure_pdf_text_quality(extracted, filename)

    if ext in {".docx", ".txt"}:
        return await asyncio.to_thread(_extract_with_unstructured, file_path)

    if ext in {".mp3", ".mp4"}:
        return await _transcribe_audio(file_path)

    raise ValueError(f"Unsupported file type: {ext}")


async def process_uploaded_files(files: list[dict]) -> str:
    combined_text = ""
    for file in files:
        original_name = file.get("originalname") or file.get("name") or "unknown-file"
        file_path = file.get("path")
        if not file_path:
            raise ValueError(f"Missing path for {original_name}")
        try:
            text = await extract_text_from_file(file_path, original_name)
        except Exception as exc:
            raise ValueError(f"Failed to parse {original_name}: {exc}") from exc

        combined_text += f"\n\n--- Document: {original_name} ---\n{text}\n"

    return combined_text

File: backend/src/utils/documentParser.js
const fs = require('fs');
const path = require('path');
const pdfParseLib = require('pdf-parse');
const mammoth = require('mammoth');
const { OpenAI } = require('openai');
const { isPageIndexEnabled, extractPdfWithPageIndex } = require('./pageIndexClient');

const openai = process.env.OPENAI_API_KEY
    ? new OpenAI({ apiKey: process.env.OPENAI_API_KEY })
    : null;

async function extractTextFromFile(filePath, filename) {
    const ext = path.extname(filename).toLowerCase();

    if (ext === '.txt') {
        return fs.readFileSync(filePath, 'utf8');
    }

    if (ext === '.pdf') {
        if (isPageIndexEnabled()) {
            try {
                const pageIndexResult = await extractPdfWithPageIndex(filePath);
                if (pageIndexResult?.text?.trim()) {
                    return pageIndexResult.text;
                }
            } catch (err) {
                console.warn(`PageIndex fallback to local parser for ${filename}: ${err.message}`);
            }
        }

        const dataBuffer = fs.readFileSync(filePath);
        // pdf-parse v2 exports PDFParse class; v1 exports a function.
        if (typeof pdfParseLib === 'function') {
            const data = await pdfParseLib(dataBuffer);
            return data.text;
        }

        if (typeof pdfParseLib.PDFParse === 'function') {
            const parser = new pdfParseLib.PDFParse({ data: dataBuffer });
            try {
                const data = await parser.getText();
                return data.text || '';
            } finally {
                await parser.destroy();
            }
        }

        throw new Error('Unsupported pdf-parse module shape');
    }

    if (ext === '.docx') {
        const result = await mammoth.extractRawText({ path: filePath });
        return result.value;
    }

    if (ext === '.mp3' || ext === '.mp4') {
        try {
            if (!openai) {
                throw new Error('OPENAI_API_KEY not configured for audio transcription');
            }
            // Using OpenAI Whisper API for audio transcription
            const transcription = await openai.audio.transcriptions.create({
                file: fs.createReadStream(filePath),
                model: 'whisper-1',
            });
            return transcription.text;
        } catch (e) {
            console.error('Audio transcription error:', e);
            throw new Error('Failed to transcribe audio file');
        }
    }

    throw new Error(`Unsupported file type: ${ext}`);
}

async function processUploadedFiles(files) {
    let combinedText = '';

    for (const file of files) {
        const storedName = file.filename || file.storedName;
        const originalName = file.originalname || file.name || 'unknown-file';
        const localPath = file.path || path.join(__dirname, '../../uploads', storedName || originalName);
        try {
            const text = await extractTextFromFile(localPath, originalName);
            combinedText += `\\n\\n--- Document: ${originalName} ---\\n${text}\\n`;
        } catch (e) {
            console.error(`Failed to parse ${originalName}`, e);
      combinedText += `\\n\\n--- Document: ${originalName} (Failed to parse) ---\\n`;
    }
  }
  
  return combinedText;
}

module.exports = { processUploadedFiles };
